### PR TITLE
Improve Multi perf on SQL state stores with 1 op only

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
When Multi is invoked with a single operation, on SQL state stores (Postgres v1/v2, CockroachDB, MySQL, SQLite, MS SQL Server) we can omit starting a transaction, as a single operation is going to be executed atomically regardless.

This should offer some non-insignificant perf benefits thanks to:

- Reducing the number of round-trips to the DB from 3 to 1
- Reducing locks used inside the DB server

### CODE REVIEW

File: common/component/postgresql/v1/postgresql.go [hunk 1]
```diff
@@ -438,29 +438,40 @@ func (p *PostgreSQL) doDelete(parentCtx context.Context, db pginterfaces.DBQueri
...
+	// If there's only 1 operation, skip starting a transaction
+	switch len(request.Operations) {
+	case 0:
+		return nil
+	case 1:
+		return p.execMultiOperation(parentCtx, request.Operations[0], p.db)
+	default:
+		_, err := pgtransactions.ExecuteInTransaction[struct{}](parentCtx, p.logger, p.db, p.metadata.Timeout, func(ctx context.Context, tx pgx.Tx) (res struct{}, err error) {
+			for _, op := range request.Operations {
+				err = p.execMultiOperation(parentCtx, op, tx)
```
Comment (elena-kolevska) [inline common/component/postgresql/v1/postgresql.go:454]: Should we use `ctx` instead of `parentContext` here, just for correctness' sake?

Comment (ItalyPaleAle) [reply common/component/postgresql/v1/postgresql.go:454]:
```suggestion
				err = p.execMultiOperation(ctx, op, tx)
```

File: state/postgresql/v2/postgresql.go [hunk 2]
```diff
@@ -468,29 +468,40 @@ func (p *PostgreSQL) doDelete(parentCtx context.Context, db pginterfaces.DBQueri
...
+	// If there's only 1 operation, skip starting a transaction
+	switch len(request.Operations) {
+	case 0:
+		return nil
+	case 1:
+		return p.execMultiOperation(parentCtx, request.Operations[0], p.db)
+	default:
+		_, err := pgtransactions.ExecuteInTransaction[struct{}](parentCtx, p.logger, p.db, p.metadata.Timeout, func(ctx context.Context, tx pgx.Tx) (res struct{}, err error) {
+			for _, op := range request.Operations {
+				err = p.execMultiOperation(parentCtx, op, tx)
```
Comment (ItalyPaleAle) [inline state/postgresql/v2/postgresql.go:484]:
```suggestion
				err = p.execMultiOperation(ctx, op, tx)
```

File: state/mysql/mysql.go [hunk 3]
```diff
@@ -768,40 +768,50 @@ func readRow(row interface{ Scan(dest ...any) error }) (key string, value []byte
...
-			if err != nil {
-				return err
+	// If there's only 1 operation, skip starting a transaction
+	switch len(request.Operations) {
+	case 0:
+		return nil
+	case 1:
+		return m.execMultiOperation(ctx, request.Operations[0], m.db)
+	default:
+		tx, err := m.db.Begin()
```
Comment (elena-kolevska) [inline state/mysql/mysql.go:785]: We should probably use `BeginTx` here like we do in sqlite.

```suggestion
		tx, err := m.db.BeginTx(ctx, nil)
```

**Review by elena-kolevska — COMMENTED** [review]
LGTM. Love that we care about saving unnecessary roundtrips.

File: state/mysql/mysql.go (hunk 3 above)
Comment (ItalyPaleAle) [reply state/mysql/mysql.go:785]: Possibly a good idea, but I'd like to not make this change in this PR and maybe do a follow-up PR

`Begin` and `BeginTx` have some difference in behavior, specifically (unlike in the pgx driver for Postgres) the context controls the lifecycle of the entire transaction. See: https://pkg.go.dev/database/sql#DB.BeginTx

**Review by yaron2 — APPROVED** [review]

**Pull request merged**